src/Distortion.cpp

```diff
@@ -162,93 +83,144 @@ std::tuple<double, double> distortionFunction(double ux, double uy,
- */
-std::tuple<double, double> removeDistortion(double inFocalPlaneX, double inFocalPlaneY,
-  const double opticalDistCoef[3], double tolerance) {
-  double rr = inFocalPlaneX * inFocalPlaneX + inFocalPlaneY * inFocalPlaneY;
-  std::tuple<double, double> undistortedPoint(inFocalPlaneX, inFocalPlaneY);
-
-  if (rr > tolerance)
-  {
-    double dr = opticalDistCoef[0] + (rr * (opticalDistCoef[1] + rr * opticalDistCoef[2]));
-    undistortedPoint = std::make_tuple(inFocalPlaneX * (1.0 - dr), inFocalPlaneY * (1.0 - dr));
+  for (int i = 0; i < 10; xPointer++, yPointer++, i++) {
+    dx = dx + f[i] * opticalDistCoeffs[xPointer];
+    dy = dy + f[i] * opticalDistCoeffs[yPointer];
   }
-
-  return undistortedPoint;
 }
 
-/**
- * @description Compute undistorted focal plane coordinate given a distorted
- * focal plane coordinate. This method works by iteratively adding distortion
- * until the new distorted point, r, undistorts to within a tolerance of the
- * original point, rp.
- *
- * @param inFocalPlaneX Distorted x
- * @param inFocalPlaneY Distorted y
- * @param opticalDistCoef Distortion coefficients
- * @param desiredPrecision Convergence precision
- * @param tolerance Tolerance of r^2
- *
- * @returns undistortedPoint Newly adjusted focal plane coordinates as an x, y tuple
- */
-std::tuple<double, double> invertDistortion(double inFocalPlaneX, double inFocalPlaneY,
-  const double opticalDistCoef[3], double desiredPrecision, double tolerance) {
-  double rp2 = (inFocalPlaneX * inFocalPlaneX) +
-               (inFocalPlaneY * inFocalPlaneY);
-  std::tuple<double, double> undistortedPoint;
-
-  if (rp2 > tolerance) {
-    double rp = sqrt(rp2);
-    // Compute first fractional distortion using rp
-    double drOverR = opticalDistCoef[0]
-                  + (rp2 * (opticalDistCoef[1] + (rp2 * opticalDistCoef[2])));
-    // Compute first distorted point estimate, r
-    double r = rp + (drOverR * rp);
-    double r_prev, r2_prev;
-    int iteration = 0;
-    do {
-      // Don't get in an end-less loop.  This algorithm should
-      // converge quickly.  If not then we are probably way outside
-      // of the focal plane.  Just set the distorted position to the
-      // undistorted position. Also, make sure the focal plane is less
-      // than 1km, it is unreasonable for it to grow larger than that.
-      if (iteration >= 15 || r > 1E9) {
-        drOverR = 0.0;
-        break;
+void removeDistortion(double dx, double dy, double &ux, double &uy,
+                      const std::vector<double> opticalDistCoeffs,
+                      DistortionType distortionType,
+                      double desiredPrecision) {
+  ux = dx;
+  uy = dy;
+
+  switch (distortionType) {
+    // Compute undistorted focal plane coordinate given a distorted
+    // coordinate set and the distortion coefficients
+    case RADIAL: {
+      double rr = dx * dx + dy * dy;
+      double tolerance = 1.0E-6;
```

Comment:
Just a for me question. Are these safe values to hard code in here for a tolerance?

Reply:
This could be replaced with the desired precision.